database/storage_engine.c

```diff
@@ -0,0 +1,282 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "storage_engine.h"
+#include "rrddim_mem.h"
+#ifdef ENABLE_DBENGINE
+#include "engine/rrdengineapi.h"
+#endif
+
+static void dimension_destroy_freez(RRDDIM *rd) {
+    freez(rd);
+}
+static void dimension_destroy_unmap(RRDDIM *rd) {
+    munmap(rd, rd->memsize);
+}
+static void set_destroy_freez(RRDSET *st) {
+    freez(st);
+}
+static void set_destroy_unmap(RRDSET *st) {
+    munmap(st, st->memsize);
+}
+
+// Initialize a rrdset with calloc and the provided memory mode
+static RRDSET* set_init_calloc(RRD_MEMORY_MODE mode) {
+    RRDSET* st = callocz(1, sizeof(RRDSET));
+    st->rrd_memory_mode = mode;
+    return st;
+}
+
+#define SET_INIT(MMODE) \
+static RRDSET* rrdset_init_##MMODE(const char *id, const char *fullid, const char *filename, long entries, int update_every) { \
+    return set_init_calloc(RRD_MEMORY_MODE_##MMODE); \
+}
+SET_INIT(NONE)
+SET_INIT(ALLOC)
+SET_INIT(DBENGINE)
+
+// Initialize a rrddim with calloc and the provided memory mode
+static RRDDIM* dim_init_calloc(RRDSET *st, RRD_MEMORY_MODE mode) {
+    size_t size = sizeof(RRDDIM) + (st->entries * sizeof(storage_number));
+    RRDDIM* rd = callocz(1, size);
+    rd->memsize = size;
+    rd->rrd_memory_mode = mode;
+    return rd;
+}
+
+#define DIM_INIT(MMODE) \
+static RRDDIM* rrddim_init_##MMODE(RRDSET *st, const char *id, const char *filename, collected_number multiplier, collected_number divisor, RRD_ALGORITHM algorithm) { \
+    return dim_init_calloc((st), RRD_MEMORY_MODE_##MMODE); \
+}
+DIM_INIT(NONE)
+DIM_INIT(ALLOC)
+DIM_INIT(DBENGINE)
+
+static const struct rrddim_collect_ops im_collect_ops = {
+    .init = rrddim_collect_init,
+    .store_metric = rrddim_collect_store_metric,
+    .finalize = rrddim_collect_finalize
+};
+
+static const struct rrddim_query_ops im_query_ops = {
+    .init = rrddim_query_init,
+    .next_metric = rrddim_query_next_metric,
+    .is_finished = rrddim_query_is_finished,
+    .finalize = rrddim_query_finalize,
+    .latest_time = rrddim_query_latest_time,
+    .oldest_time = rrddim_query_oldest_time
+};
+
+STORAGE_ENGINE engines[] = {
+    {
+        .id = RRD_MEMORY_MODE_NONE,
+        .name = RRD_MEMORY_MODE_NONE_NAME,
+        .api = {
+            .engine_ops = {NULL, NULL, NULL},
```

Comment:
Better to have a dummy function instead of NULLs, as it will simplify the sites that use these fields.

Reply:
Done (will be part of the third PR)